### DISCUSSION TITLE
C# Script to Delete All Measures

### DISCUSSION THREAD

**Discussion by Declan1984** [TabularEditor/TabularEditor3#377] [🙋♀️TE3 Q&A] [answered]
I'm trying to write a script that will delete all existing measures on a table.  I'm relatively new to both C# and Tabular Editor so still trying to get a handle on how to leverage scripts to automate my modeling.

Here is my code but it threw an error; can anyone share any insights on the best way to handle:

```
var delete_measures = Model.Tables["Test Model"].Measures;

foreach(var measure_item in delete_measures){
    var measure_name = measure_item.Name;
    var measure_to_delete = Model.Tables["Test Model"].Measures[measure_name];
    measure_to_delete.Delete();
}

```

**Comment by otykier** [comment, marked as answer]
In C#, you can't modify a collection while iterating it, so you have to copy the items into a new collection like so:

```csharp
var delete_measures = Model.Tables["Test Model"].Measures.ToList(); // .ToList() creates a copy of the collection

foreach(var measure in delete_measures) {
    measure.Delete(); 
    // No reason to look up the measure by name, since we already have a reference to it...
}
```

**Reply by Declan1984** [reply]
@otykier thank you for that code snippet; this scripting functionality in TE3 is amazing!  In lieu of, or until, composite models are implemented, I want to have a single primary definition script of all measures that can be run against any dataset model that pulls in data from our core set of Data Flow tables.  This script checks if the dataset it is being run in has a particular table (will expand to all tables) and if that table has measures, and if so, clear the measures.  Below this will be the code that defines the measures so the dataset would be updated from the primary script.  There are probably a lot of improvements that could be made to this, such as functionalizing the delete loops but the basic outline is there.

Now if only there were a Python wrapper for running these scripts ;-)


```
//Table Check

bool tbl_exist_Target_Table = false;

var model_tables_list = Model.Tables.ToList();

foreach(var table_obj in model_tables_list){
	string table_name = table_obj.Name;
    if(table_name == "Target Table"){
		tbl_exist_Target_Table = true;
	}
}

/*
Delete existing measures
*/

//Target Table
if(tbl_exist_Target_Table == true){
	var delete_measures_Target_Table = Model.Tables["Target Table"].Measures.ToList();
    
    int num_of_measures = delete_measures_Target_Table.Count;
	
    if(num_of_measures > 0){
        foreach(var measure_item in delete_measures_Target_Table){
            measure_item.Delete();
        }
    }
}

```
